internal/controller/pkg/revision/hook.go

```diff
@@ -215,6 +228,128 @@ func (h *ConfigurationHooks) Post(context.Context, runtime.Object, v1.PackageRev
+	return &FunctionHooks{
+		client:         client,
+		namespace:      namespace,
+		serviceAccount: serviceAccount,
+	}
+}
+
+// Pre cleans up a packaged controller and service account if the revision is
+// inactive.
+func (h *FunctionHooks) Pre(ctx context.Context, pkg runtime.Object, pr v1.PackageRevision) error {
+	fo, _ := xpkg.TryConvert(pkg, &pkgmetav1alpha1.Function{})
+	pkgFunction, ok := fo.(*pkgmetav1alpha1.Function)
+	if !ok {
+		return errors.New(errNotFunction)
+	}
+
+	// TODO(ezgidemirel): update any status fields relevant to package revisions.
+
+	// Do not clean up SA and controller if revision is not inactive.
+	if pr.GetDesiredState() != v1.PackageRevisionInactive {
+		return nil
+	}
+
+	// NOTE(hasheddan): we avoid fetching pull secrets and controller config as
+	// they aren't needed to delete Deployment, ServiceAccount, and Service.
+	s, d, _, _ := buildFunctionDeployment(pkgFunction, pr, nil, h.namespace, []corev1.LocalObjectReference{})
+	if err := h.client.Delete(ctx, d); resource.IgnoreNotFound(err) != nil {
+		return errors.Wrap(err, errDeleteFunctionDeployment)
+	}
+	if err := h.client.Delete(ctx, s); resource.IgnoreNotFound(err) != nil {
+		return errors.Wrap(err, errDeleteFunctionSA)
+	}
+
+	return nil
+}
+
+// Post creates a packaged function deployment, service account, service and secrets if the revision is active.
+func (h *FunctionHooks) Post(ctx context.Context, pkg runtime.Object, pr v1.PackageRevision) error { //nolint:gocyclo // See below
+	// TODO(ezgidemirel): Can this be refactored for less complexity?
+	po, _ := xpkg.TryConvert(pkg, &pkgmetav1alpha1.Function{})
+	pkgFunction, ok := po.(*pkgmetav1alpha1.Function)
+	if !ok {
+		return errors.New(errNotFunction)
+	}
+	if pr.GetDesiredState() != v1.PackageRevisionActive {
+		return nil
+	}
+	cc, err := h.getControllerConfig(ctx, pr)
+	if err != nil {
+		return err
+	}
+	ps, err := h.getSAPullSecrets(ctx)
+	if err != nil {
+		return err
+	}
+	s, d, svc, secSer := buildFunctionDeployment(pkgFunction, pr, cc, h.namespace, append(pr.GetPackagePullSecrets(), ps...))
+	if err := h.client.Apply(ctx, s); err != nil {
+		return errors.Wrap(err, errApplyFunctionSA)
+	}
+	owner := pr.GetOwnerReferences()
+	if err := h.client.Apply(ctx, secSer); err != nil {
+		return errors.Wrap(err, errApplyFunctionSecret)
+	}
+	if err := h.client.Apply(ctx, d); err != nil {
+		return errors.Wrap(err, errApplyFunctionDeployment)
+	}
+	if pr.GetTLSServerSecretName() != nil && pr.GetTLSClientSecretName() != nil {
+		if err := initializer.NewTLSCertificateGenerator(h.namespace, initializer.RootCACertSecretName, *pr.GetTLSServerSecretName(), *pr.GetTLSClientSecretName(), pkgFunction.Name, initializer.TLSCertificateGeneratorWithOwner(owner)).GenerateServerCertificate(ctx, h.client); err != nil {
+			return errors.Wrapf(err, "cannot generate TLS certificates for %s", pkgFunction.Name)
+		}
+	}
```

Comment:
nit: maybe switch the apply and the secret initialisation, just to make it similar to the hook for providers

Reply:
done